sklearn_numba_dpex/common/kernels.py

```diff
@@ -397,7 +429,149 @@ def partial_sum_reduction(
+        check_power_of_2(
+            n_sub_group_per_work_group := (work_group_size // sub_group_size)
+        )
+
+    else:
+        # Round work_group_size to the maximum smaller power-of-two multiple of
+        # `sub_group_size`
+        n_sub_group_per_work_group = 2 ** (
+            math.floor(math.log2(work_group_size / sub_group_size))
+        )
+        work_group_size = n_sub_group_per_work_group * sub_group_size
+
+    # Number of iteration in each execution of the kernel:
+    local_n_iterations = np.int64(math.floor(math.log2(n_sub_group_per_work_group)) - 1)
+
+    local_values_size = (n_sub_group_per_work_group, sub_group_size)
+    reduction_block_size = 2 * n_sub_group_per_work_group
+
+    # ???: how does this strategy compares to having each thread reducing N contiguous
+    # items ?
+    @dpex.kernel
+    # fmt: off
+    def partial_sum_reduction(
+        summands,    # IN        (n_rows, sum_axis_size)
+        result,      # OUT       (n_rows, math.ceil(size / (2 * reduction_block_size),)
+    ):
+
+        # fmt: on
+        # NB: This kernel only perform a partial reduction
+        group_id = dpex.get_group_id(zero_idx)
+        local_work_id = dpex.get_local_id(zero_idx)
+
+        sum_axis_size = summands.shape[zero_idx]
+        n_block_for_col = result.shape[zero_idx]
+
+        local_block_id_in_col = (group_id % n_block_for_col)
+        first_row_idx = local_block_id_in_col * reduction_block_size
+        local_col_idx = (local_work_id % sub_group_size)
+        local_row_idx = (local_work_id // sub_group_size)
+        col_idx = (
+            group_id // n_block_for_col) * sub_group_size + local_col_idx
+        augend_row_idx = first_row_idx + local_row_idx
+        addend_row_idx = augend_row_idx + n_sub_group_per_work_group
+
+        local_values = dpex.local.array(local_values_size, dtype=dtype)
+
+        # Each work item reads two value in global memory and sum it into the local
+        # memory
+        if (col_idx >= n_cols) or (augend_row_idx >= sum_axis_size):
+            local_values[local_row_idx, local_col_idx] = zero
+        elif addend_row_idx >= sum_axis_size:
+            local_values[local_row_idx, local_col_idx] = fused_unary_func_(
+                summands[augend_row_idx, col_idx])
+        else:
+            local_values[local_row_idx, local_col_idx] = fused_unary_func_(
+                summands[augend_row_idx, col_idx]) + fused_unary_func_(
+                    summands[addend_row_idx, col_idx])
+
+        dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+        current_n_nows = n_sub_group_per_work_group
+        for i in range(local_n_iterations):
+            # We discard half of the remaining active subgroups at each iteration
+            current_n_nows = current_n_nows // two_as_a_long
+            if (local_row_idx < current_n_nows) and (
+                    col_idx < n_cols) and (
+                        first_row_idx + local_row_idx + current_n_nows) < sum_axis_size:
```

Comment:
Is this really black formatted? I would have expected something like the following or similar:

```suggestion
            if (
                (local_row_idx < current_n_nows) and
                (col_idx < n_cols) and
                (first_row_idx + local_row_idx + current_n_nows) < sum_axis_size
            ):
                
```

Reply:
not sure about the position of the `and`s though.

Reply:
Although you're right your suggestion is prettier (and in fact @jjerphan suggested the same in the PR for the cleaning of the `axis1` sum, and I forgot to apply it for `axis0` again :sob:  ), the former _is_ black formatted, I don't think there's an issue with `fmt` tags here. Answer applies to other similar comments :grin:

Reply:
> the former is black formatted, I don't think there's an issue with fmt tags here. Answer applies to other similar comments 😁

I checked and black is disabled in this block. I don't know how to make the `fmt on/off` things to work to disable black only inside the prototype of the function unfortunately.

Reply:
I think I found a solution: removing the `# fmt: off` / `# fmt: on` lines and instead appending a `# fmt: skip` at the end of the line with the `):` of the function prototype seems to work. Let me try to update this file accordingly.

Reply:
Hum, it works when using the command line manually but then the pre-commit command fails... Let me investigate.

Reply:
The version of black in the pre-commit config is probably too old to handle this pattern. I will re-run experiments with both versions after lunch.

Reply:
Thanks for the investigation !!